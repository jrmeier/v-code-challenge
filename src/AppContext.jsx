import { useState, createContext, useEffect } from 'react';
import { addShoppingItem, fetchShoppingList, editShoppingItem, deleteShoppingItem } from './api';


export const updateLocalStorageWithItems = (items) => {
    localStorage.setItem('items', JSON.stringify(items))
    }

export const getItemsFromLocalStorage = () => {
    return JSON.parse(localStorage.getItem('items'))
}

const defaultAppContext = {
    items: [],
    setItems: Function
}

export const AppContext = createContext(defaultAppContext);


export function AppContextProvider(props) {
    const [items, setItems] = useState(defaultAppContext.items);
    const [loading, setLoading] = useState(false);

    const shoppingListId = 1

    const addItem = async (newItem) => {
        // add the item to the shopping list
        const newItemWithId = await addShoppingItem(shoppingListId, newItem)
        const newItems = [...items, newItemWithId];
        setItems(newItems)
        
        // updateLocalStorageWithItems(newItems)
    }

    const editItem = (newItem) => {
        editShoppingItem(shoppingListId, newItem)
        // updateLocalStorageWithItems(newItems)
        const itemIndex = items.findIndex((item) => item.id === newItem.id)
        const newItems = [...items]
        newItems[itemIndex] = newItem
        setItems(newItems)
    }

    const loadItems = async () => {
        setLoading(true)
        const shoppingList = await fetchShoppingList(shoppingListId)
        setItems(shoppingList.shoppingListItems)
        setLoading(false)
    }

    const deleteItem = (itemId) => {
        deleteShoppingItem(shoppingListId, itemId)
        const newItems = items.filter((item) => item.id !== itemId);
        setItems(newItems);
    }


  useEffect(() => {
    console.log('useEffect called in app context');
    const fetchData = async () => {
      loadItems();
    };

    fetchData();
  }, []); // eslint-disable-line react-hooks/exhaustive-deps

    const value = {
        items,
        setItems,
        addItem,
        editItem,
        loadItems,
        loading,
        deleteItem
    }

    return (
        <AppContext.Provider value={value}>
            {props.children}
        </AppContext.Provider>
    )
}
