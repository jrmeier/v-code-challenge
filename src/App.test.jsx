import React from 'react';
import { render, screen } from '@testing-library/react';
import {
  describe, it, expect, vi,
} from 'vitest';
import { ThemeProvider } from '@mui/material/styles';

import { AppContext, AppContextProvider } from './AppContext';
import AppWithProvider, { AppNoProvider, theme } from './App';

vi.mock(AppContext, () => ({
  items: [],
  setItems: () => [],
  loadItems: () => [],
  loading: false,

  addItem: () => [],
  editItem: () => [],
  deleteItem: () => [],
}));

vi.mock('./api/localStorageHelpers', () => ({
  updateLocalStorageWithItems: () => [],
  getItemsFromLocalStorage: () => [],
}));

describe('AppNoProvider', () => {
  it('renders the AppBar with the correct text', () => {
    render(
      <ThemeProvider theme={theme}>
        <AppNoProvider />
      </ThemeProvider>,
    );
    const appbarText = screen.getByText('SHOPPING LIST');
    expect(appbarText).toBeTruthy();
  });

  it('renders the LoadingScreen when loading is true', () => {
    render(
      <ThemeProvider theme={theme}>
        <AppNoProvider />
      </ThemeProvider>,
    );
    const loadingScreen = screen.findByTestId('loadingSpinner');
    expect(loadingScreen).toBeTruthy();
  });

  it('renders the ShoppingList when loading is false', () => {
    render(
      <ThemeProvider theme={theme}>
        <AppNoProvider />
      </ThemeProvider>,
    );
    const shoppingList = screen.findByTestId('shoppingList');
    expect(shoppingList).toBeTruthy();
  });
});

describe('AppWithProvider', () => {
  it('renders the AppBar with the correct text', () => {
    const rendered = render(
      <ThemeProvider theme={theme}>
        <AppWithProvider />
      </ThemeProvider>,
    );

    expect(rendered).toMatchSnapshot();
  });
});
