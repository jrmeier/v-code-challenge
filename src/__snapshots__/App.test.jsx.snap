// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AppWithProvider > renders the AppBar with the correct text 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="App"
      >
        <header
          class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionStatic css-hip9hq-MuiPaper-root-MuiAppBar-root"
          style="box-shadow: none;"
        >
          <div
            class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-hyum1k-MuiToolbar-root"
          >
            <h6
              class="MuiTypography-root MuiTypography-h6 css-q06dfn-MuiTypography-root"
            >
              SHOPPING LIST
            </h6>
          </div>
        </header>
        <div
          class="MuiBox-root css-dlo6ih"
        >
          <h5
            class="MuiTypography-root MuiTypography-h5 css-19pxqrz-MuiTypography-root"
          >
            Your shopping list is empty :( 
          </h5>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-eacpwv-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
             
            Add your first item
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
    <div>
      <div
        class="App"
      >
        <header
          class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionStatic css-hip9hq-MuiPaper-root-MuiAppBar-root"
          style="box-shadow: none;"
        >
          <div
            class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-hyum1k-MuiToolbar-root"
          >
            <h6
              class="MuiTypography-root MuiTypography-h6 css-q06dfn-MuiTypography-root"
            >
              SHOPPING LIST
            </h6>
          </div>
        </header>
        <div
          class="MuiBox-root css-dlo6ih"
        >
          <h5
            class="MuiTypography-root MuiTypography-h5 css-19pxqrz-MuiTypography-root"
          >
            Your shopping list is empty :( 
          </h5>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-eacpwv-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
             
            Add your first item
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
    <div>
      <div
        class="App"
      >
        <header
          class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionStatic css-hip9hq-MuiPaper-root-MuiAppBar-root"
          style="box-shadow: none;"
        >
          <div
            class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-hyum1k-MuiToolbar-root"
          >
            <h6
              class="MuiTypography-root MuiTypography-h6 css-q06dfn-MuiTypography-root"
            >
              SHOPPING LIST
            </h6>
          </div>
        </header>
        <div
          class="MuiBox-root css-dlo6ih"
        >
          <h5
            class="MuiTypography-root MuiTypography-h5 css-19pxqrz-MuiTypography-root"
          >
            Your shopping list is empty :( 
          </h5>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-eacpwv-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
             
            Add your first item
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
    <div>
      <div
        class="App"
      >
        <header
          class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionStatic css-hip9hq-MuiPaper-root-MuiAppBar-root"
          style="box-shadow: none;"
        >
          <div
            class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-hyum1k-MuiToolbar-root"
          >
            <h6
              class="MuiTypography-root MuiTypography-h6 css-q06dfn-MuiTypography-root"
            >
              SHOPPING LIST
            </h6>
          </div>
        </header>
        <div
          class="MuiBox-root css-dlo6ih"
        >
          <h5
            class="MuiTypography-root MuiTypography-h5 css-19pxqrz-MuiTypography-root"
          >
            Your shopping list is empty :( 
          </h5>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-eacpwv-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
             
            Add your first item
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="App"
    >
      <header
        class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionStatic css-hip9hq-MuiPaper-root-MuiAppBar-root"
        style="box-shadow: none;"
      >
        <div
          class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-hyum1k-MuiToolbar-root"
        >
          <h6
            class="MuiTypography-root MuiTypography-h6 css-q06dfn-MuiTypography-root"
          >
            SHOPPING LIST
          </h6>
        </div>
      </header>
      <div
        class="MuiBox-root css-dlo6ih"
      >
        <h5
          class="MuiTypography-root MuiTypography-h5 css-19pxqrz-MuiTypography-root"
        >
          Your shopping list is empty :( 
        </h5>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-eacpwv-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
           
          Add your first item
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
